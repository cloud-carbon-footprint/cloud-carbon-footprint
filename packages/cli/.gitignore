dist
*.csv
*.cache.json
estimates.cache.*.json
*.snap
.env
.env.production
!aws_input.test.csv
!gcp_input.test.csv
!azure_input.test.csv
!incorrect_input.test.csv
!on_premise_data_input.test.csv