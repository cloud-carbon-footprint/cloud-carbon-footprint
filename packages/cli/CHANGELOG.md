# @cloud-carbon-footprint/cli

## 1.1.0

### Minor Changes

- 61332214: Adds CLI command for creating a lookup table to be used in ETL pipelines
- 52237bc6: Adds additional dashboard for viewing cloud provider recommendations including refactoring filters for reusability

  For updating the create-app templates, please refer to the following [commit](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/commit/d1f9a94ea88e6f9210a781e16df18b9f64a0b03d).

### Patch Changes

- Updated dependencies [f3569daa]
- Updated dependencies [61332214]
- Updated dependencies [52237bc6]
  - @cloud-carbon-footprint/app@0.3.0
  - @cloud-carbon-footprint/common@1.2.0

## 1.0.2

### Patch Changes

- @cloud-carbon-footprint/app@0.2.1

## 1.0.1

### Patch Changes

- 52a8b3c1: Adds support for specifying recommendation target for AWS via API parameter

  Please refer to this [commit](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/commit/404c77796ae838766dc8007e18daee2c7526f6ed) to update create app templates.

- Updated dependencies [f75cf08f]
- Updated dependencies [52a8b3c1]
  - @cloud-carbon-footprint/app@0.2.0
  - @cloud-carbon-footprint/common@1.1.0

## 1.0.0

### Major Changes

- 91ed3d75: Update variables GCP_BILLING_ACCOUNT_ID/GCP_BILLING_ACCOUNT_NAME to be GCP_BILLING_PROJECT_ID/GCP_BILLING_PROJECT_NAME

  This is a major update for packages: api, cli, common. Please refer to this [commit](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/commit/aab00ea5a395d94ba5ce1424ffa33abbafd7ed58) for create-app template updates as they are breaking changes.

### Patch Changes

- bbceed8b: Updates dependencies: @types/fs-extra, typescript, husky and @types/node
- Updated dependencies [13f39ac7]
- Updated dependencies [91ed3d75]
- Updated dependencies [72fc2752]
- Updated dependencies [e76d5fdd]
- Updated dependencies [bbceed8b]
  - @cloud-carbon-footprint/app@0.1.0
  - @cloud-carbon-footprint/common@1.0.0

## 0.4.2

### Patch Changes

- 3b42775b: Updates @types/node to 15.12.4
- Updated dependencies [e93d31ec]
- Updated dependencies [3b42775b]
- Updated dependencies [53366130]
  - @cloud-carbon-footprint/common@0.0.3
  - @cloud-carbon-footprint/app@0.0.4

## 0.4.1

### Patch Changes

- Updated dependencies [8e24c32b]
  - @cloud-carbon-footprint/common@0.0.2
  - @cloud-carbon-footprint/app@0.0.3

## 0.4.0

### Minor Changes

- ababb826: Extracts two new packages app and common to avoid circular dependancies and make it easier to extract cloud provider packages
- e84a4c7a: Extract logic into the new packages: app, common, gcp, aws, azure:

  There are many files that have been updated/extracted.
  In order to update create-app templates, refer to the follow [commit](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/commit/8c6aaed52e9f3949e134852986d50362aad3367a).

  The following changes were made to,
  'packages/create-app/templates/default-app/packages/client/tsconfig.json':

  ```diff
      // ...
          "skipLibCheck": true,
          "esModuleInterop": true,
          "allowSyntheticDefaultImports": true,
  -        "strict": true,
          "forceConsistentCasingInFileNames": true,
          "module": "esnext",
          "moduleResolution": "node",
          "resolveJsonModule": true,
  -        "isolatedModules": true,
          "noEmit": true,
          "jsx": "react-jsx",
  -        "noFallthroughCasesInSwitch": true
  +        "noFallthroughCasesInSwitch": true,
  +        "strict": false,
  +        "isolatedModules": true
        },
  -      "include": [
  -        "src",
  -        "node_modules/apexcharts/types/apexcharts.d.ts"
  -      ]
  +      "include": ["src"]
      }
      // ...
  ```

  Additionally, the following dependencies have been updated and should also be updated in their respective template package.json file:

  - @cloud-carbon-footprint root package.json:
    - "@types/fs-extra": "^9.0.11"
    - "concurrently": "^6.2.0"
    - "marked": ">=2.0.5"
  - @cloud-carbon-footprint/api and @cloud-carbon-footprint/cli:
    - "dotenv": "^10.0.0"
    - "@cloud-carbon-footprint/app": Can be added with `yarn up @cloud-carbon-footprint/app`
    - "@cloud-carbon-footprint/common": Can be added with `yarn up @cloud-carbon-footprint/common`
  - @cloud-carbon-footprint/client:
    - "dotenv": "^10.0.0"
    - "@testing-library/react-hooks": "^7.0.0"
    - "concurrently": "^6.2.0"
    - "@cloud-carbon-footprint/common": Can be added with `yarn up @cloud-carbon-footprint/common`

### Patch Changes

- c7fa7db0: Updates dependencies to the latest
- Updated dependencies [f9fbcb4c]
- Updated dependencies [c7fa7db0]
- Updated dependencies [ababb826]
- Updated dependencies [e84a4c7a]
  - @cloud-carbon-footprint/core@0.8.0

## 0.3.3

### Patch Changes

- b63d8a67: The default `aws-sdk` dependency was bumped to `"^2.890.0"`,
- Updated dependencies [56bb6da6]
- Updated dependencies [68365cbf]
- Updated dependencies [b63d8a67]
- Updated dependencies [8df5703b]
- Updated dependencies [3e2f876d]
- Updated dependencies [3abe3dca]
- Updated dependencies [29f48e7c]
- Updated dependencies [370c509d]
- Updated dependencies [7d523b59]
  - @cloud-carbon-footprint/core@0.7.0

## 0.3.2

### Patch Changes

- 81dfde6c: Improves .env.template files with links to cofiguration glossary
- b0151b7d: Improves GCP guided-install prompts in the CLI and create-app packages
- Updated dependencies [8b4c992d]
- Updated dependencies [216ea2ec]
  - @cloud-carbon-footprint/core@0.6.0

## 0.3.1

### Patch Changes

- 8175b41d: updates packages with readmes and metadata
- Updated dependencies [8175b41d]
- Updated dependencies [0ccce96b]
  - @cloud-carbon-footprint/core@0.5.0

## 0.3.0

### Minor Changes

- 23d841e: Adds support for guided install command in the cli package.

### Patch Changes

- Updated dependencies [6c620db]
- Updated dependencies [f3d4c8a]
- Updated dependencies [c5f28fe]
- Updated dependencies [906f14e]
- Updated dependencies [bb82d7b]
- Updated dependencies [033a504]
- Updated dependencies [9b10f3b]
- Updated dependencies [82aeb1e]
  - @cloud-carbon-footprint/core@0.4.0

## 0.2.1

### Patch Changes

- a0ac965: Fixes failing tests when running them in a different timezone
- Updated dependencies [beabadf]
- Updated dependencies [e2ac070]
- Updated dependencies [29e9897]
- Updated dependencies [75176d7]
- Updated dependencies [ce0f249]
- Updated dependencies [29e9897]
- Updated dependencies [a0ac965]
  - @cloud-carbon-footprint/core@0.3.0

## 0.2.0

### Minor Changes

- 96cb023: Adds support for Microsoft Azure, as well as updates to the microsite.

### Patch Changes

- Updated dependencies [96cb023]
  - @cloud-carbon-footprint/core@0.2.0

## 0.1.0

### Minor Changes

- 3b3bec3: Update packages to have base version
