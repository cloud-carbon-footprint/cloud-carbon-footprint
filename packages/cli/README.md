# @cloud-carbon-footprint/cli

This package provides a Command Line Interface as an entrypoint to get cloud energy and carbon emissions.

## Installation

Install the package via npm or Yarn:

```sh
$ npm install --save @cloud-carbon-footprint/cli
```

or

```sh
$ yarn add @cloud-carbon-footprint/cli
```

## Development

### Running Locally

Please refer to the [documentation](https://www.cloudcarbonfootprint.org/docs/running-the-cli) on how to run the cli locally.

### Guided Install

```
yarn guided-install
```

This will install dependencies for all packages, then guide you through setting up credentials and environment variables to analyze the footprint of your AWS, GCP, or Azure account. Then:

```
yarn start
```

If you have any problems with the guided install, you can instead [configure your cloud provider environment variables manually](https://www.cloudcarbonfootprint.org/docs/introduction#connecting-your-data).

## Documentation

- [Cloud Carbon Footprint Readme](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/trunk/README.md)
- [Cloud Carbon Footprint Documentation](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/tree/trunk/microsite/docs/README.md)

## License

Â© 2021 ThoughtWorks, Inc.
