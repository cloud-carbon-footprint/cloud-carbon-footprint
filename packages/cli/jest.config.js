/*
 * Â© 2021 Thoughtworks, Inc.
 */

// eslint-disable-next-line @typescript-eslint/no-var-requires
const baseConfig = require('../../jest.base.config')

module.exports = {
  ...baseConfig,
  coverageThreshold: {
    global: {
      statements: 97,
      branches: 87,
      functions: 96,
      lines: 97,
    },
  },
  coveragePathIgnorePatterns: [
    '<rootDir>/src/CliPrompts.ts',
    '<rootDir>/src/CreateLookupTable/createAzureInput.ts',
    '<rootDir>/src/__tests__/fixtures',
  ],
  testPathIgnorePatterns: ['<rootDir>/src/__tests__/fixtures'],
  modulePathIgnorePatterns: ['index.ts'],
}
