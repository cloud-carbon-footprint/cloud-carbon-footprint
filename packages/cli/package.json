{
  "name": "@cloud-carbon-footprint/cli",
  "version": "1.1.0",
  "license": "Apache-2.0",
  "description": "Command Line Interface as an entrypoint to get cloud energy and carbon emissions.",
  "main": "src/index.ts",
  "publishConfig": {
    "main": "dist/index.js"
  },
  "homepage": "https://www.cloudcarbonfootprint.org/",
  "repository": {
    "type": "git",
    "url": "https://github.com/cloud-carbon-footprint/cloud-carbon-footprint",
    "directory": "packages/cli"
  },
  "keywords": [
    "thoughtworks",
    "cloud",
    "sustainability",
    "climate",
    "carbon-emissions",
    "carbon-footprint"
  ],
  "scripts": {
    "precommit": "lint-staged --no-stash && yarn test --coverage --runInBand --bail",
    "test": "jest",
    "type-check": "tsc --noEmit",
    "sonar:scan": "sonar-scanner",
    "test:watch": "yarn test --coverage --watchAll",
    "build": "rimraf dist && yarn build:tsc",
    "build:tsc": "tsc --build tsconfig.json",
    "prepack": "ts-node ../../scripts/prepack.ts",
    "postpack": "ts-node ../../scripts/postpack.ts",
    "build:watch": "onchange 'src/**/*.ts' -- yarn build",
    "view:coverage": "serve coverage/lcov-report",
    "clean": "rimraf dist && rimraf coverage && rimraf logs",
    "lint": "eslint '*/**/*.ts' --quiet",
    "lint:fix": "eslint '*/**/*.ts' --quiet --fix",
    "start": "ts-node src/index.ts",
    "guided-install": "ts-node src/GuidedInstall/index.ts",
    "create-lookup-table": "ts-node src/CreateLookupTable/index.ts"
  },
  "devDependencies": {
    "@types/jest": "^27.0.1",
    "@types/jest-when": "^2.7.2",
    "@types/node": "^16.0.1",
    "@types/source-map-support": "^0.5.3",
    "@typescript-eslint/eslint-plugin": "^4.24.0",
    "@typescript-eslint/parser": "^4.24.0",
    "aws-sdk-mock": "^5.1.0",
    "eslint": "^7.26.0",
    "eslint-config-prettier": "^8.3.0",
    "eslint-plugin-prettier": "^3.4.0",
    "eslint-plugin-unused-imports": "^1.1.1",
    "jest": "26.6.0",
    "jest-when": "^3.3.1",
    "lint-staged": "^11.0.1",
    "onchange": "^7.1.0",
    "prettier": "^2.3.0",
    "rimraf": "^3.0.2",
    "source-map-support": "^0.5.19",
    "ts-jest": "^26.5.6",
    "ts-node": "^10.2.1",
    "ts-node-dev": "^1.1.6",
    "typescript": "^4.3.5"
  },
  "dependencies": {
    "@cloud-carbon-footprint/app": "0.3.0",
    "@cloud-carbon-footprint/common": "^1.2.0",
    "@types/cli-table": "^0.3.0",
    "@types/prompts": "^2.0.12",
    "@types/ramda": "^0.27.40",
    "aws-sdk": "^2.910.0",
    "cli-table": "^0.3.6",
    "commander": "^8.0.0",
    "csv-writer": "^1.6.0",
    "csvtojson": "^2.0.10",
    "dotenv": "^10.0.0",
    "fs-extra": "^10.0.0",
    "moment": "^2.29.1",
    "ramda": "^0.27.1",
    "typed-prompts": "^1.5.0"
  },
  "lint-staged": {
    "*.{js,ts}": [
      "yarn lint",
      "prettier --write --ignore-unknown --config ../../.prettierrc.json"
    ],
    "*.{md}": [
      "prettier --write --ignore-unknown --config ../../.prettierrc.json"
    ]
  }
}
