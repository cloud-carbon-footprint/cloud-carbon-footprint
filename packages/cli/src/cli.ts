/*
 * Â© 2021 ThoughtWorks, Inc.
 */

import commander from 'commander'
import {
  App,
  CreateValidRequest,
  EstimationResult,
} from '@cloud-carbon-footprint/app'
import * as process from 'process'
import EmissionsByDayAndServiceTable from './EmissionsByDayAndServiceTable'
import EmissionsByServiceTable from './EmissionsByServiceTable'
import EmissionsByDayTable from './EmissionsByDayTable'
import CliPrompts from './CliPrompts'
import { exportToCSV } from './CSV'
import moment from 'moment'
import path from 'path'

export default async function cli(argv: string[] = process.argv) {
  const program = new commander.Command()
  program.storeOptionsAsProperties(false)

  program
    .option('-s, --startDate <string>', 'Start date in ISO format')
    .option('-e, --endDate <string>', 'End date in ISO format')
    .option('-r, --region <string>', 'AWS region to analyze')
    .option(
      '-g, --groupBy <string>',
      'Group results by day or service. Default is day.',
    )
    .option('-i, --interactive', 'Use interactive CLI prompts')
    .option(
      '-f, --format <string>',
      'How to format the results [table, csv]. Default is table.',
    )

  program.parse(argv)

  let startDate, endDate
  let region
  let groupBy: string
  let format: string

  if (program.opts().interactive) {
    ;[startDate, endDate, region, groupBy, format] = await CliPrompts()
  } else {
    const programOptions = program.opts()
    startDate = programOptions.startDate
    endDate = programOptions.endDate
    region = programOptions.region
    groupBy = programOptions.groupBy
    format = programOptions.format
  }
  const estimationRequest = CreateValidRequest({ startDate, endDate, region })
  const { table, colWidths } = await new App()
    .getCostAndEstimates(estimationRequest)
    .then((estimations: EstimationResult[]) => {
      if (groupBy === 'service') {
        return EmissionsByServiceTable(estimations)
      }
      if (groupBy === 'day') {
        return EmissionsByDayTable(estimations)
      }
      return EmissionsByDayAndServiceTable(estimations)
    })

  if (format === 'csv') {
    const filePath = path.join(
      process.cwd(),
      `results-${moment().utc().format('YYYY-MM-DD-HH:mm:ss')}.csv`,
    )
    exportToCSV(table, filePath)
    return `File saved to: ${filePath}`
  } else {
    return table
      .map((row: string[]) =>
        row.reduce(
          (acc, data, col) => acc + `| ${data}`.padEnd(colWidths[col]),
          '',
        ),
      )
      .join('\n')
  }
}
