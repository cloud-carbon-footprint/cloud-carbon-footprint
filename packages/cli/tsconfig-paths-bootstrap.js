/* eslint-disable */
const tsConfig = require('./tsconfig.json')
/* eslint-disable */
const tsConfigPaths = require('tsconfig-paths')

let { baseUrl, paths } = tsConfig.compilerOptions
for (path in paths) {
  paths[path][0] = paths[path][0].replace('src', 'dist').replace('.ts', '.js')
}

tsConfigPaths.register({ baseUrl, paths })
