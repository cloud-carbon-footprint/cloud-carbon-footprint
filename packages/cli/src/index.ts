/*
 * Â© 2021 ThoughtWorks, Inc.
 */

import cli from './cli'
import * as console from 'console'

cli().then(console.log)
