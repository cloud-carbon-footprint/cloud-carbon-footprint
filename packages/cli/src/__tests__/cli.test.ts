/*
 * Â© 2021 Thoughtworks, Inc.
 */

import AWSMock from 'aws-sdk-mock'
import AWS from 'aws-sdk'
import {
  EstimationRequestValidationError,
  EstimationResult,
  GroupBy,
  configLoader,
} from '@cloud-carbon-footprint/common'
import {
  mockAwsCloudWatchGetMetricData,
  mockAwsCloudWatchGetQueryResultsForLambda,
  mockAwsCostExplorerGetCostAndUsage,
  mockAwsCostExplorerGetCostAndUsageResponse,
} from './fixtures/awsMockFunctions'
import { lambdaMockGetCostResponse } from './fixtures/costexplorer.fixtures'
import {
  mockCpuUtilizationTimeSeries,
  mockVCPUTimeSeries,
} from './fixtures/cloudmonitoring.fixtures'
import cli from '../cli'
import { MongoDbCacheManager } from '@cloud-carbon-footprint/app'

let mockWarn: jest.SpyInstance
const mockListTimeSeries = jest.fn()

jest.mock('@google-cloud/monitoring', () => {
  return {
    MetricServiceClient: jest.fn().mockImplementation(() => {
      return {
        listTimeSeries: mockListTimeSeries,
        projectPath: jest
          .fn()
          .mockReturnValue('projects/cloud-carbon-footprint'),
        getProjectId: jest.fn().mockResolvedValue('cloud-carbon-footprint'),
      }
    }),
  }
})

const mockGetCostAndEstimates = jest.fn()
const mockGetFilterData = jest.fn()
const mockGetEmissionsFactors = jest.fn()

jest.mock('@cloud-carbon-footprint/app', () => ({
  ...(jest.requireActual('@cloud-carbon-footprint/app') as Record<
    string,
    unknown
  >),
  App: jest.fn().mockImplementation(() => {
    return {
      getCostAndEstimates: mockGetCostAndEstimates,
      getEmissionsFactors: mockGetEmissionsFactors,
      getFilterData: mockGetFilterData,
    }
  }),
  cache: jest.fn(),
}))

jest.mock('@cloud-carbon-footprint/common', () => ({
  ...(jest.requireActual('@cloud-carbon-footprint/common') as Record<
    string,
    unknown
  >),
  Logger: jest.fn().mockReturnValue({
    debug: jest.fn(),
    info: jest.fn(),
    error: jest.fn(),
    warn: jest.fn(),
  }),
  configLoader: jest.fn().mockImplementation(() => {
    return {
      AWS: {
        accounts: [{ id: '12345678', name: 'test account' }],
        NAME: 'AWS',
        CURRENT_REGIONS: ['us-east-1', 'us-east-2'],
        CURRENT_SERVICES: [
          {
            key: 'ebs',
            name: 'EBS',
          },
          {
            key: 's3',
            name: 'S3',
          },
          {
            key: 'ec2',
            name: 'EC2',
          },
          {
            key: 'elasticache',
            name: 'ElastiCache',
          },
          {
            key: 'rds',
            name: 'RDS',
          },
          {
            key: 'lambda',
            name: 'Lambda',
          },
        ],
        authentication: {
          mode: 'GCP',
          options: {
            targetRoleName: 'test-target',
            proxyAccountId: 'test-account-id',
            proxyRoleName: 'test-role-name',
          },
        },
      },
      GCP: {
        projects: [{ id: 'test-project', name: 'test project' }],
        NAME: 'GCP',
        CURRENT_REGIONS: ['us-east1'],
        CURRENT_SERVICES: [
          {
            key: 'computeEngine',
            name: 'ComputeEngine',
          },
        ],
      },
      LOGGING_MODE: 'test',
    }
  }),
}))

beforeAll(() => {
  AWSMock.setSDKInstance(AWS)
})

beforeEach(() => {
  mockWarn = jest.spyOn(console, 'warn').mockImplementation()
})

afterEach(() => {
  AWSMock.restore()
})

describe('cli', () => {
  const start = '2020-07-01'
  const end = '2020-07-07'
  const grouping = 'day'
  const rawRequest = [
    'executable',
    'file',
    '--startDate',
    start,
    '--endDate',
    end,
    '--groupBy',
    grouping,
  ]

  beforeEach(() => {
    const expectedResponse: EstimationResult[] = []
    mockGetCostAndEstimates.mockResolvedValueOnce(expectedResponse)
  })

  afterEach(() => {
    jest.restoreAllMocks()
  })

  describe('request parameters', () => {
    const expectedValidatedParams = {
      startDate: new Date(start),
      endDate: new Date(end),
      groupBy: GroupBy.day,
      ignoreCache: false,
    }

    it('passes all required request parameters to App request', async () => {
      await cli([...rawRequest])
      expect(mockGetCostAndEstimates).toHaveBeenCalledWith(
        expectedValidatedParams,
      )
    })
  })

  describe('ebs, s3, ec3, elasticache, rds', () => {
    beforeEach(() => {
      mockAwsCloudWatchGetMetricData()
      mockAwsCostExplorerGetCostAndUsage()
      mockListTimeSeries
        .mockResolvedValueOnce([mockCpuUtilizationTimeSeries, {}, {}])
        .mockResolvedValueOnce([mockVCPUTimeSeries, {}, {}])
    })

    test('ebs, s3, ec2, elasticache, rds, grouped by day and service', async () => {
      const result = await cli(rawRequest)

      expect(result).toMatchSnapshot()
    })

    test('ebs, s3, ec2, elasticache, rds grouped by service', async () => {
      const result = await cli([...rawRequest, '--groupBy', 'service'])

      expect(result).toMatchSnapshot()
    })

    test('ebs, s3, ec2, elasticache, rds, grouped by day', async () => {
      const result = await cli([...rawRequest, '--groupBy', 'day'])

      expect(result).toMatchSnapshot()
    })
  })

  describe('lambda', () => {
    beforeEach(() => {
      mockAwsCloudWatchGetQueryResultsForLambda()
      mockAwsCostExplorerGetCostAndUsageResponse(lambdaMockGetCostResponse)
    })

    it('lambda estimates', async () => {
      const result = await cli([...rawRequest, '--groupBy', 'dayAndService'])
      expect(result).toMatchSnapshot()
    })

    it('lambda cost', async () => {
      const result = await cli([...rawRequest, '--groupBy', 'service'])
      expect(result).toMatchSnapshot()
    })
  })

  describe('parameter validation', () => {
    describe('given: invalid dates', () => {
      it('throws an estimation validation error', async () => {
        // setup
        const start = '2020-06-16'
        const end = '2020-06-15'
        const command = [
          'executable',
          'file',
          '--startDate',
          start,
          '--endDate',
          end,
        ]

        //assert
        await expect(() => {
          return cli([...command])
        }).rejects.toThrow(EstimationRequestValidationError)
      })
    })

    describe('given: invalid table grouping parameter', () => {
      it('throws an estimation request validation error', async () => {
        const grouping = 'year'
        const command = [
          'executable',
          'file',
          '--startDate',
          start,
          '--endDate',
          end,
          '--groupBy',
          grouping,
        ]

        const validGroupByParams = ['day', 'dayAndService', 'service']
        const errorMessage = `GroupBy param is incorrect. Please specify one of the following grouping methods: ${validGroupByParams.join(
          ' | ',
        )}`

        await expect(() => {
          return cli([...command])
        }).rejects.toThrow(new EstimationRequestValidationError(errorMessage))
      })

      it('assigns a default value of "day" along with a warning when not provided', async () => {
        const command = [
          'executable',
          'file',
          '--startDate',
          start,
          '--endDate',
          end,
        ]

        await cli([...command])
        expect(mockWarn).toHaveBeenCalledWith(
          'GroupBy parameter not specified, adopting "day" as the default.',
        )
      })
    })
  })

  describe('database connection', () => {
    const mockCreateDbConnection = jest.fn()
    const mockClient: any = { close: jest.fn() }
    beforeEach(() => {
      MongoDbCacheManager.createDbConnection = mockCreateDbConnection
      MongoDbCacheManager.mongoClient = mockClient
      ;(configLoader as jest.Mock).mockReturnValue({
        ...configLoader(),
        CACHE_MODE: 'MONGODB',
      })
    })

    it('should initialize MongoDB client when it is configured as the cache mode', async () => {
      const consoleSpy = jest.spyOn(console, 'log').mockImplementation()

      await cli([...rawRequest])
      expect(mockCreateDbConnection).toHaveBeenCalledTimes(1)
      expect(mockClient.close).toHaveBeenCalledTimes(1)
      expect(consoleSpy).toHaveBeenCalledWith('MongoDB connection closed')
    })
  })
})
