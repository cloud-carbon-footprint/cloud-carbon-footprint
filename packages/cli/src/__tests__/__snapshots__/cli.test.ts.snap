// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli ebs, s3, ec3, elasticache, rds ebs, s3, ec2, elasticache, rds grouped by service 1`] = `
"| Service      | kilowatt hours    | metric tons CO2e Emissions| Cost              
| EBS          | 6.08              | 0.002602               | $9.23             
| S3           | 0.08              | 0.000036               | $12.00            
| EC2          | 39.48             | 0.016895               | $74.00            
| ElastiCache  | 24.06             | 0.010295               | $12.00            
| RDS          | 464.46            | 0.198774               | $18.47            
| ComputeEngine| 5.78              | 0.002892               | $0.00             
| Total        | 539.94            | 0.231494               | $125.70           "
`;

exports[`cli ebs, s3, ec3, elasticache, rds ebs, s3, ec2, elasticache, rds, grouped by day 1`] = `
"| Date         | kilowatt hours    | metric tons CO2e Emissions
| 2020-07-01   | 60.56             | 0.026335               
| 2020-07-02   | 479.37            | 0.205159               
| Total        | 539.94            | 0.231494               "
`;

exports[`cli ebs, s3, ec3, elasticache, rds ebs, s3, ec2, elasticache, rds, grouped by day and service 1`] = `
"| Date (UTC)   | EBS kilowatt hours| EBS metric tons CO2e Emissions| S3 kilowatt hours | S3 metric tons CO2e Emissions| EC2 kilowatt hours| EC2 metric tons CO2e Emissions| ElastiCache kilowatt hours| ElastiCache metric tons CO2e Emissions| RDS kilowatt hours| RDS metric tons CO2e Emissions| ComputeEngine kilowatt hours| ComputeEngine metric tons CO2e Emissions| SUM kilowatt hours| SUM metric tons CO2e Emissions
| 2020-07-01   | 2.03              | 0.000867               | 0.08              | 0.000036               | 39.48             | 0.016895               | 5.12              | 0.002193               | 8.07              | 0.003452               | 5.78              | 0.002892               | 60.56             | 0.026335               
| 2020-07-02   | 4.05              | 0.001735               | 0.00              | 0.000000               | 0.00              | 0.000000               | 18.93             | 0.008102               | 456.39            | 0.195322               | 0.00              | 0.000000               | 479.37            | 0.205159               
| Total        | 6.08              | 0.002602               | 0.08              | 0.000036               | 39.48             | 0.016895               | 24.06             | 0.010295               | 464.46            | 0.198774               | 5.78              | 0.002892               | 539.94            | 0.231494               "
`;

exports[`cli lambda lambda cost 1`] = `
"| Service      | kilowatt hours    | metric tons CO2e Emissions| Cost              
| EBS          | 6.08              | 0.002602               | $9.23             
| S3           | 0.08              | 0.000036               | $12.00            
| EC2          | 39.48             | 0.016895               | $74.00            
| ElastiCache  | 24.06             | 0.010295               | $12.00            
| RDS          | 464.46            | 0.198774               | $18.47            
| ComputeEngine| 5.78              | 0.002892               | $0.00             
| Total        | 539.94            | 0.231494               | $125.70           "
`;

exports[`cli lambda lambda estimates 1`] = `
"| Date (UTC)   | EBS kilowatt hours| EBS metric tons CO2e Emissions| S3 kilowatt hours | S3 metric tons CO2e Emissions| EC2 kilowatt hours| EC2 metric tons CO2e Emissions| ElastiCache kilowatt hours| ElastiCache metric tons CO2e Emissions| RDS kilowatt hours| RDS metric tons CO2e Emissions| ComputeEngine kilowatt hours| ComputeEngine metric tons CO2e Emissions| SUM kilowatt hours| SUM metric tons CO2e Emissions
| 2020-07-01   | 2.03              | 0.000867               | 0.08              | 0.000036               | 39.48             | 0.016895               | 5.12              | 0.002193               | 8.07              | 0.003452               | 5.78              | 0.002892               | 60.56             | 0.026335               
| 2020-07-02   | 4.05              | 0.001735               | 0.00              | 0.000000               | 0.00              | 0.000000               | 18.93             | 0.008102               | 456.39            | 0.195322               | 0.00              | 0.000000               | 479.37            | 0.205159               
| Total        | 6.08              | 0.002602               | 0.08              | 0.000036               | 39.48             | 0.016895               | 24.06             | 0.010295               | 464.46            | 0.198774               | 5.78              | 0.002892               | 539.94            | 0.231494               "
`;
